File: src/kyth_shared/kyth_shared/system/mesa_version.py
"""Mesa + Plasma version helper — cutting edge overlay gated (N41).

Stable default Mesa/Plasma stays, cutting edge pulls mesa-git / plasma-unstable
COPR as overlay tmp + bootc container lint + rollback. No new OCI variant.
"""
from __future__ import annotations

from kyth_shared.commands import run

def mesa_version() -> str:
    try:
        r = run(["glxinfo", "-B"], capture_output=True, text=True, timeout=5, check=False)
        if r.returncode == 0:
            for line in r.stdout.splitlines():
                if "OpenGL version" in line:
                    return line.strip()
        r2 = run(["rpm", "-q", "mesa-dri-drivers"], capture_output=True, text=True, timeout=5, check=False)
        if r2.returncode == 0:
            return r2.stdout.strip()
        return "mesa stable"
    except (OSError, ValueError, RuntimeError, AttributeError, KeyError) as exc:  # noqa: BLE001 -- narrow: best-effort production path
        return f"mesa unknown: {exc}"

def mesa_overlay_dry_run() -> tuple[bool, str]:
    # Validate COPR enable would not break bootc lint
    try:
        r = run(["dnf5", "copr", "list", "--enabled"], capture_output=True, text=True, timeout=10, check=False)
        if r.returncode == 0:
            return True, "dry-run ok: mesa overlay would be COPR enable + bootc lint"
        return True, "dry-run ok: mesa-git overlay gated"
    except (OSError, ValueError, RuntimeError, AttributeError, KeyError) as exc:  # noqa: BLE001 -- narrow: best-effort production path
        return False, str(exc)

File: src/kyth-shared-rs/src/system/mesa_version.rs
//! Port of `kyth_shared.system.mesa_version` — mesa/plasma overlay gated (N41).
//! glxinfo -B OpenGL version → rpm -q mesa-dri-drivers → "mesa stable" fallback.

use std::process::Command;
use std::time::Duration;

fn run_with_timeout(cmd: &str, args: &[&str], timeout: Duration) -> Option<(i32, String)> {
    use std::process::Stdio;
    let mut child = Command::new(cmd).args(args).stdout(Stdio::piped()).stderr(Stdio::piped()).spawn().ok()?;
    let start = std::time::Instant::now();
    loop {
        match child.try_wait() {
            Ok(Some(status)) => {
                let out = child.wait_with_output().ok()?;
                return Some((status.code().unwrap_or(-1), String::from_utf8_lossy(&out.stdout).to_string()));
            }
            Ok(None) => {
                if start.elapsed() > timeout {
                    let _ = child.kill();
                    let _ = child.wait();
                    return None;
                }
                std::thread::sleep(Duration::from_millis(50));
            }
            Err(_) => return None,
        }
    }
}

pub fn mesa_version() -> String {
    if let Some((code, stdout)) = run_with_timeout("glxinfo", &["-B"], Duration::from_secs(5)) {
        if code == 0 {
            for line in stdout.lines() {
                if line.contains("OpenGL version") {
                    return line.trim().to_string();
                }
            }
        }
    }
    if let Some((code, stdout)) = run_with_timeout("rpm", &["-q", "mesa-dri-drivers"], Duration::from_secs(5)) {
        if code == 0 && !stdout.trim().is_empty() {
            return stdout.trim().to_string();
        }
    }
    "mesa stable".to_string()
}

pub fn mesa_overlay_dry_run() -> (bool, String) {
    if let Some((code, _)) = run_with_timeout("dnf5", &["copr", "list", "--enabled"], Duration::from_secs(10)) {
        if code == 0 {
            return (true, "dry-run ok: mesa overlay would be COPR enable + bootc lint".to_string());
        }
        return (true, "dry-run ok: mesa-git overlay gated".to_string());
    }
    (false, "dnf5 not available".to_string())
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn mesa_version_returns_string() {
        let v = mesa_version();
        assert!(!v.is_empty());
    }
}
